### DISCUSSION TITLE
Run background service on internal transition ?

### DISCUSSION THREAD

**Discussion by lbdremy** [statelyai/xstate#3895] [Q&A]
Hello,

I am wondering how I should proceed to correctly do that using xstate.
I have a state machine with multiple states, 3 states have an internal transition, this internal transition trigger an action but this action is now _asynchronous_ , so I should use a service there but invokation of service in transition is not supported it seems.
So I think I'm probably not thinking this right.
Here are my constraints:
- I want to keep the same state when this transition is triggered (that's the reason why I'm thinking internal transition)
- I want to update the machine context at the beginning of the asynchronous operation and also when it succeed or when it fails 

Other questions:
- How to set things up to be able cancel the asynchronous operation based on the state changes that happen in the mean time ?

Thanks a lot for your help 🙏

**Comment by Andarist** [comment]
1. use `invoke` on the parent state and just send start/stop/restart events to it as necessary
2. or use a parallel region like:
```ts
type: 'parallel',
states: {
  asyncJob: {
    initial: 'idle',
    states: {
      idle: {},
      running: {
        invoke: { src: 'task', onDone: { target: 'idle', actions: raise({ type: 'COMPLETED' }) } },
        entry: assign({ running: true }),
        exit: assign({ running: false })
      },
      on: {
        RESTART: 'running',
        STOP: 'idle'
      }
    },
  },
  foo: { /* communicate with the other state as needed */ }
}
```
